design-reference/architecture/VersionAndGo/readme.md

```diff
@@ -0,0 +1,113 @@
+ - `Version` - all versioning schemes should use [SemVer](https://semver.org/) or variant. 
+
+## Version and Go Setup Description
+Our design starts after the initial Cobalt setup, where an individual template is cloned into an organizations source control (ref: [Setup Doc](https://github.com/microsoft/cobalt/blob/master/docs/GETTING_STARTED_APP_DEV.md)). An application development team is ready to start deploying their code to infrastructure built in Azure. Typically they would make a request to the engineering team. The Eng team would provide the App team with a link to the latest infra configuration files for a specific template. The App team downloads this tar/zip file and extracts the folder/files.
+
+```
+infra/
+ |_ Application.tf
+ |_ Application.yaml
+```
+
+These files should be derived from the template folder in dev repo and cloned into the eng repo. Proposed folder path/location is:
+
+```
+infra/
+  |_template
+      |_template_name
+          |_tests
+          |_config
+          |   |_ application.tf
+          |   |_ application.yml
+          |_template files...
+```
+
+ - `Appliation.tf` is the artifact that points to the template in the eng repo. The application team would need to fill this file out with the requested configuration. Example:
+
+ ```
+ module "az-isolated-service-single-region" {
+   source = "git://dev.azure.com/project/repo?ver=1.0.0//infra/template/az-isolated-service-single-region/"
+   resource_group_location = "eastus"
+   name = "az-simple"
+   deployment_targets = [{
+     app_name = "cobalt-backend-api",
+     image_name = "msftcse/az-service-single-region",
+     image_release_tag_prefix = "release"
+   }]
+   acr_build_git_source_url = "https://github.com/erikschlegel/echo-server.git"
+}
+ ```
+
+ - `Application.yaml` - is the artifact that points to the azure-pipeline.yml file in the eng repo that contains the build/test/deploy pipeline.
+
+This folder/files should be checked into the root of the app repo. The Eng team would then configure a new pipeline in the app teams Azure Devops project to point to these files and deploy the requested infrastructure.
+
+> Note: We should strive to continue to use a single set of pipeline files for all pipelines referenced in this document.
+
+## Version and Go Eng Pipeline Description
+
+This pipeline will produce an artifact that is stored in a location that would be accessible by both the engineering team and the app team. It should follow or use the same pipeline as the development team and app team pipelines. This code that publishes a new version could be controlled by a feature flag as global variable in the eng pipeline. This could also use a SemVer global variable and only create a new artifact when this has been incremented. Would could also leave this as a "manual" process. Manual only means that the pipeline doesn't create the artifact. We could use a script or the likes to do this.
```

Comment:
Note: not entierly sure I understand this section, so my questions may be off-base a bit. 

> This pipeline will produce an artifact that is stored in a location that would be accessible by both the engineering team and the app team

Are you saying that this pipeline essentially produces a new "version" (artifact) of the terraform source module that can be consumed by both teams? I.e., publish to terraform registry, create a tag in git, etc...?

> It should follow or use the same pipeline as the development team and app team pipelines

Can you elaborate on why? The infrastructure focused pipelines follow a pattern of terraform commands & tests that execute in a given environment. The pipeline described here seems to be a version publish. Unclear on how much re-use there is between these steps - can you help me understand?

> controlled by a feature flag

Will this amount to the following? If so, I'm not seeing the benefit of using a single pipeline definition.

```
if shouldPublishNewVersion == True:
    < run publish workflow >
else:
    < deploy to $STAGE >
```